Cubical/Categories/Constructions/Hyperdoctrine.agda

```diff
@@ -0,0 +1,63 @@
+{-# OPTIONS --safe #-}
+
+module Cubical.Categories.Constructions.Hyperdoctrine where
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Categories.Category
+open import Cubical.Categories.Functor.Base
+open import Cubical.Categories.NaturalTransformation.Base
+
+open import Cubical.Categories.Instances.PosetCat.Base
+open import Cubical.Categories.Instances.PosetCat.Monotone
+open import Cubical.Categories.Instances.PosetCat.MonotoneAdjoint
+open import UMP
+
+private
+  variable
+    ℓ ℓ' : Level
+
+open CartesianCategory
+open Category
+open MonFunAdj
+
+record Hyperdoctrine ℓ ℓ' : Type (ℓ-suc (ℓ-max ℓ ℓ')) where
+  field
+    cc : CartesianCategory ℓ ℓ'
+    func : Functor ((cc .cat) ^op) (PosetCatAdjCycl {ℓ} {ℓ'})
+    {-
+    Beck-Chevalley condition
+    Expresses the naturality of the left/right adjoints of the a × _ projection:
+    For f ∈ cc [b , c], we have the following commuting square:
+
+                              func(π₂)
+                 func(a × c) ⟵--------- func(c)
+                   |         ---------→   |
+      func(1 × f)  |         left-adj     | func(f)
+                   |                      |
+                   ↓          func(π₂)    ↓
+                 func(a × b) ←--------- func(b)
+                             ---------→
+                             left-adj
+
+    TODO: Can probably be more cleanly expressed as a bifunctor
```

Comment:
I'll have to think about if there's a clean formulation of this property. I would guess the Kleisli category of the Env comonad is involved